java/vector/src/main/java/org/apache/arrow/vector/BaseVariableWidthVector.java

```diff
@@ -169,6 +169,42 @@ public void setInitialCapacity(int valueCount) {
     offsetAllocationSizeInBytes = (valueCount + 1) * OFFSET_WIDTH;
   }
 
+  /**
+   * Sets the desired value capacity for the vector. This function doesn't
+   * allocate any memory for the vector.
+   * @param valueCount desired number of elements in the vector
+   * @param density average number of bytes per variable width element
+   */
+  public void setInitialCapacity(int valueCount, double density) {
+    final long size = (long) (valueCount * density);
+    if (size < 1) {
+      throw new IllegalArgumentException("With the provided density and value count, potential capacity of the data buffer is 0");
+    }
+    if (size > MAX_ALLOCATION_SIZE) {
+      throw new OversizedAllocationException("Requested amount of memory is more than max allowed");
+    }
+    valueAllocationSizeInBytes = (int) size;
```

Comment:
Do we need to round this up to nearest 2^n?

Reply:
It happens during actual allocation. setInitialCapacity() doesn't allocate but just records the state for amount of memory we want to allocate.